File: app/src/main/java/go/planner/plannergo/AssignmentActivity.java
package go.planner.plannergo;

import android.app.DatePickerDialog;
import android.app.TimePickerDialog;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.annotation.ColorRes;
import android.support.annotation.IdRes;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.view.menu.ActionMenuItemView;
import android.support.v7.widget.ActionMenuView;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.AutoCompleteTextView;
import android.widget.CheckBox;
import android.widget.DatePicker;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.Spinner;
import android.widget.TimePicker;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Locale;

import static java.util.Calendar.DAY_OF_MONTH;
import static java.util.Calendar.HOUR_OF_DAY;
import static java.util.Calendar.MINUTE;
import static java.util.Calendar.MONTH;
import static java.util.Calendar.YEAR;

public abstract class AssignmentActivity extends AppCompatActivity {
    //Reads and writes involve this object
    NewAssignment assignment;
//    Calendar notifyDate = Calendar.getInstance();
//    Calendar notifyDate2 = Calendar.getInstance();

    //Settings file
    SharedPreferences prefs;

    //All the fucking views all up in one big list because Kotlin's nicer isn't it
    Toolbar toolbar;
    EditText hw_title;
    AutoCompleteTextView hw_class;
    EditText hw_due_date;
    EditText hw_due_time;
    EditText hw_description;
    Spinner hw_type;
    CheckBox is_priority;
    LinearLayout custom_notification;
    CheckBox enable_custom_notification;
    LinearLayout reminder_display;
    EditText n_days;
    EditText n_time;
    LinearLayout custom_notification_2;
    CheckBox extra_notification;
    LinearLayout extra_reminder_display;
    EditText r2_date;
    EditText r2_time;


    //Listeners
    TimePickerDialog dueTimePickerDialog;
    DatePickerDialog dueDatePickerDialog;
    TimePickerDialog notifyTimePickerDialog;
    TimePickerDialog notifyExtraTimePickerDialog;

    SimpleDateFormat dateFormat = new SimpleDateFormat("EEE, MMM dd, yyyy", Locale.US);
    SimpleDateFormat timeFormat = new SimpleDateFormat("h:mm a".toLowerCase(), Locale.US);

    /**
     * implementation should include super.onCreate(savedInstanceState;
     * and setContentView(R.layout.activity_assignment);
     *
     * @param savedInstanceState for restoring state
     */
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_assignment);

        prefs = PreferenceManager.getDefaultSharedPreferences(this);

        Intent thisIntent = getIntent();

        assignment = FileIO.getAssignment(thisIntent.getExtras() != null ? thisIntent.getExtras().getLong("uniqueID", -1L) : -1L);
        Log.v("AssignmentActivity", "assignment=" + assignment);

        connectViews();

        int layoutID = android.R.layout.simple_dropdown_item_1line;
        ArrayList<String> classArrayList = FileIO.classNames;
        String[] classes = classArrayList.toArray(new String[classArrayList.size()]);
        ArrayAdapter<String> classAdapter = new ArrayAdapter<>(this, layoutID, classes);
        hw_class.setAdapter(classAdapter);
        hw_class.setThreshold(0);



        ArrayList<String> typesArrayList = FileIO.types;
        String[] types = typesArrayList.toArray(new String[typesArrayList.size()]);
        ArrayAdapter<String> typesAdapter = new ArrayAdapter<>(this, layoutID, types);
//
//        ArrayAdapter<CharSequence> adapter = ArrayAdapter.createFromResource(this,
//                R.array.assignment_types_array, android.R.layout.simple_spinner_item);
//        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        hw_type.setAdapter(typesAdapter);

        manageVisibility();
        setUpListeners();
        initViews();
        initToolbar();
    }

    void initToolbar(){
        toolbar.setBackgroundColor(ColorPicker.getColorAssignment());
        toolbar.setTitleTextColor(ColorPicker.getColorAssignmentText());
        Drawable menuIcon;
        if (ColorPicker.getColorAssignmentText() == Color.BLACK) {
            toolbar.setNavigationIcon(R.drawable.ic_arrow_back_black_24dp);
            menuIcon = ContextCompat.getDrawable(getApplicationContext(),
                    R.drawable.ic_more_vert_black_24dp);
        } else {
            toolbar.setNavigationIcon(R.drawable.ic_arrow_back_white_24dp);
            menuIcon = ContextCompat.getDrawable(getApplicationContext(),
                    R.drawable.ic_more_vert_white_24dp);
        }
        toolbar.setOverflowIcon(menuIcon);
        getWindow().setStatusBarColor(ColorPicker.getColorAssignmentAccent());
        setSupportActionBar(toolbar);
    }


    /**
     * Determines which views should be visible when the activity starts.
     */
    abstract void manageVisibility();

    void connectViews() {
        toolbar = findViewById(R.id.toolbar);
        hw_title = findViewById(R.id.hw_title);
        hw_class = findViewById(R.id.hw_class);
        hw_due_date = findViewById(R.id.hw_due_date);
        hw_due_time = findViewById(R.id.hw_due_time);
        hw_description = findViewById(R.id.hw_description);
        hw_type = findViewById(R.id.hw_type);
        is_priority = findViewById(R.id.is_priority);
        custom_notification = findViewById(R.id.custom_notification);
        enable_custom_notification = findViewById(R.id.enable_custom_notification);
        reminder_display = findViewById(R.id.reminder_display);
        n_days = findViewById(R.id.n_days);
        n_time = findViewById(R.id.n_time);
        custom_notification_2 = findViewById(R.id.custom_notification_2);
        extra_notification = findViewById(R.id.extra_notification);
        extra_reminder_display = findViewById(R.id.extra_reminder_display);
        r2_date = findViewById(R.id.r2_date);
        r2_time = findViewById(R.id.r2_time);
    }

    abstract void initViews();

    abstract void setUpListeners();

    public static void setToolbarMenuItemTextColor(final Toolbar toolbar, final @ColorRes int color, @IdRes final int resId) {
        if (toolbar != null) {
            for (int i = 0; i < toolbar.getChildCount(); i++) {
                final View view = toolbar.getChildAt(i);
                if (view instanceof ActionMenuView) {
                    final ActionMenuView actionMenuView = (ActionMenuView) view;
                    // view children are accessible only after layout-ing
                    actionMenuView.post(new Runnable() {
                        @Override
                        public void run() {
                            for (int j = 0; j < actionMenuView.getChildCount(); j++) {
                                final View innerView = actionMenuView.getChildAt(j);
                                if (innerView instanceof ActionMenuItemView) {
                                    final ActionMenuItemView itemView = (ActionMenuItemView) innerView;
                                    if (resId == itemView.getId()) {
                                        itemView.setTextColor(ContextCompat.getColor(toolbar.getContext(), color));
                                    }
                                }
                            }
                        }
                    });
                }
            }
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.new_assignment_menu, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()){
            case R.id.save_assignment:
                Log.v("NewAssignmentActivity", "save button pressed");
                saveAssignment();
            case android.R.id.home:
                navigateUpTo(new Intent(this, MainActivity.class));
                return true;
                default:
                    return super.onOptionsItemSelected(item);


        }
    }

    /**
     * creates dialog to modify a date field
     *
     * @param dateToModify 0 for dueDate, 1 for notificationDate1, 2 for NotificationDate2
     * @param view         field to modify
     * @return TimePickerDialog
     */
    TimePickerDialog createTimePicker(final int dateToModify, final EditText view) {
        final Calendar calendar;
        switch (dateToModify) {
            default:
            case 0:
                calendar = (assignment.dueDate == null) ?
                        Calendar.getInstance() : assignment.dueDate;
                break;
            case 1:
                calendar = (assignment.notificationDate1 == null) ?
                        Calendar.getInstance() : assignment.notificationDate1;
                break;
            case 2:
                calendar = (assignment.notificationDate2 == null) ?
                        Calendar.getInstance() : assignment.notificationDate2;
                break;
        }

        return new TimePickerDialog(this, new TimePickerDialog.OnTimeSetListener() {
            @Override
            public void onTimeSet(TimePicker timePicker, int hourOfDay, int minute) {
                calendar.set(HOUR_OF_DAY, hourOfDay);
                calendar.set(MINUTE, minute);
                view.setText(timeFormat.format(calendar.getTime()));
                switch (dateToModify) {
                    default:
                    case 0:
                        assignment.dueDate = calendar;
                        break;
                    case 1:
                        assignment.notificationDate1 = calendar;
                        break;
                    case 2:
                        assignment.notificationDate2 = calendar;
                }
            }
        }, calendar.get(HOUR_OF_DAY), calendar.get(MINUTE), false);

    }

    /**
     * reads and writes to date
     * outputs changes to view
     */
    DatePickerDialog createDatePicker(final Calendar date, final EditText view) {
        return new DatePickerDialog(this, new DatePickerDialog.OnDateSetListener() {
            @Override
            public void onDateSet(DatePicker datePicker, int year, int month, int dayOfMonth) {
                date.set(YEAR, year);
                date.set(MONTH, month);
                date.set(DAY_OF_MONTH, dayOfMonth);
                view.setText(dateFormat.format(date.getTime()));
            }
        }, date.get(YEAR), date.get(MONTH), date.get(DAY_OF_MONTH));
    }

    abstract public void toggleCustomNotification(View view);

    abstract public void toggleExtraNotification(View view);

    public void togglePriorityAssignment(View view) {
        if (assignment.priority == 1) {
            assignment.priority = 0;
            is_priority.setChecked(false);
        } else {
            assignment.priority = 1;
            is_priority.setChecked(true);
        }
    }


    abstract void saveAssignment();

}

File: app/src/main/java/go/planner/plannergo/AssignmentActivity.kt
package go.planner.plannergo

import android.app.AlertDialog
import android.app.DatePickerDialog
import android.app.TimePickerDialog
import android.content.Intent
import android.content.SharedPreferences
import android.content.res.ColorStateList
import android.graphics.Color
import android.os.Bundle
import android.preference.PreferenceManager
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import androidx.core.graphics.ColorUtils
import kotlinx.android.synthetic.main.activity_assignment.*
import java.text.SimpleDateFormat
import java.util.*
import java.util.Calendar.*

abstract class AssignmentActivity : AppCompatActivity(), ColorSchemeActivity {
    private val tag = "AssignmentActivity"

    //Keeps old data in case edits were accidental
    private lateinit var oldAssignment: Assignment
    //Reads and writes involve this object
    internal lateinit var mAssignment: Assignment

    //Settings file
    internal lateinit var prefs: SharedPreferences
    internal lateinit var colorScheme: ColorScheme
    private var schemeSet = false

    //Listeners
    internal var dueTimePickerDialog: TimePickerDialog? = null
    internal var dueDatePickerDialog: DatePickerDialog? = null

    internal var dateFormat = SimpleDateFormat("EEE, MMM dd, yyyy", Locale.US)
    internal var timeFormat = SimpleDateFormat("h:mm a".toLowerCase(Locale.US), Locale.US)

    private lateinit var typeSpinner: Spinner
    private var layoutID = 0

    /**
     * implementation should include super.onCreate(savedInstanceState;
     * and setContentView(R.layout.activity_assignment);
     *
     * @param savedInstanceState for restoring state
     */
    override fun onCreate(savedInstanceState: Bundle?) {
        prefs = PreferenceManager.getDefaultSharedPreferences(this)
        setColorScheme()
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_assignment)


        val thisIntent = intent

        mAssignment = FileIO.getAssignment(if (thisIntent.extras != null) thisIntent.extras!!.getLong("uniqueID", -1L) else -1L)
        Log.v("AssignmentActivity", "mAssignment=$mAssignment")
        oldAssignment = Assignment(mAssignment)

        //Set up autocomplete for class field
        val classArrayList = FileIO.classNames
        val classes = classArrayList.toTypedArray()

        layoutID = if (colorScheme == ColorScheme.SCHEME_DARK)
            R.layout.spinner_item_dark
        else
            android.R.layout.simple_dropdown_item_1line
        val classAdapter = ArrayAdapter(this, layoutID, classes)


        hw_class.setAdapter(classAdapter)
        hw_class.threshold = 0

        updateTypeSpinner()

        manageVisibility()
        setUpListeners()
        initViews()
        setSupportActionBar(toolbar)
    }

    override fun onResume() {
        checkForColorSchemeUpdate()
        super.onResume()
    }

    /**
     * Fills spinner with options from FileIO.types
     */
    private fun updateTypeSpinner() {
        val typesArrayList = FileIO.types
        val types = typesArrayList.toTypedArray()

        layoutID = if (colorScheme == ColorScheme.SCHEME_DARK)
            R.layout.spinner_item_dark
        else
            android.R.layout.simple_dropdown_item_1line

        val typesAdapter = ArrayAdapter(this, layoutID, types)
        typeSpinner = findViewById(R.id.hw_type)
        typeSpinner.adapter = typesAdapter
    }


    /**
     * Determines which views should be visible when the activity starts.
     */
    internal abstract fun manageVisibility()

    internal abstract fun initViews()

    internal abstract fun setUpListeners()

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.new_assignment_menu, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onPrepareOptionsMenu(menu: Menu?): Boolean {
        menu?.getItem(0)?.icon = colorScheme.getDrawable(this, Field.AS_APP_BAR_OPT)
        return super.onPrepareOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.save_assignment -> {
                Log.v("NewAssignmentActivity", "save button pressed")
                saveAssignment()
                navigateUpTo(Intent(this, MainActivity::class.java))
                true
            }
            android.R.id.home -> {
                saveCheckAndNavigateUp()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onBackPressed() {
        saveCheckAndNavigateUp()
    }

    private fun saveCheckAndNavigateUp() {
        Log.d(tag, "oldAssignment == assignment?${oldAssignment.compareFields(getAssignment())}")
        if (oldAssignment.compareFields(getAssignment()))
            navigateUpTo(Intent(this, MainActivity::class.java))
        else {
            //TODO: add don't ask me again option
            val style = if (colorScheme == ColorScheme.SCHEME_DARK)
                R.style.DarkDialogTheme
            else
                R.style.LightDialogTheme
            AlertDialog.Builder(this, style)
                    .setTitle(R.string.do_not_save)
                    .setMessage(R.string.changes_wont_be_saved)
                    .setNegativeButton(R.string.leave) { _, _ ->
                        navigateUpTo(Intent(this, MainActivity::class.java))
                    }
                    .setNeutralButton(R.string.leave_and_save) { _, _ ->
                        saveAssignment()
                        navigateUpTo(Intent(this, MainActivity::class.java))
                    }
                    .setPositiveButton(R.string.stay, null)
                    .create().show()
        }
    }

    /**
     * creates dialog to modify a date field
     *
     * @param dateToModify 0 for dueDate, 1 for notificationDate1, 2 for NotificationDate2
     * @param view         field to modify
     * @return TimePickerDialog
     */
    internal fun createTimePicker(dateToModify: Int, view: EditText): TimePickerDialog {
        val calendar: Calendar
        when (dateToModify) {
            0 -> calendar =
                    if (mAssignment.dueDate == null)
                        getInstance()
                    else
                        mAssignment.dueDate
            1 -> calendar =
                    if (mAssignment.notificationDate1 == null)
                        getInstance()
                    else
                        mAssignment.notificationDate1
            2 -> calendar =
                    if (mAssignment.notificationDate2 == null)
                        getInstance()
                    else
                        mAssignment.notificationDate2
            else -> calendar =
                    if (mAssignment.dueDate == null)
                        getInstance()
                    else
                        mAssignment.dueDate
        }

        return TimePickerDialog(this, TimePickerDialog.OnTimeSetListener { _, hourOfDay, minute ->
            calendar.set(HOUR_OF_DAY, hourOfDay)
            calendar.set(MINUTE, minute)
            view.setText(timeFormat.format(calendar.time))
            when (dateToModify) {
                0 -> mAssignment.dueDate = calendar
                1 -> mAssignment.notificationDate1 = calendar
                2 -> mAssignment.notificationDate2 = calendar
                else -> mAssignment.dueDate = calendar
            }
        }, calendar.get(HOUR_OF_DAY), calendar.get(MINUTE), false)

    }

    /**
     * reads and writes to date
     * outputs changes to view
     */
    internal fun createDatePicker(date: Calendar, view: EditText): DatePickerDialog {
        return DatePickerDialog(this, DatePickerDialog.OnDateSetListener { _, year, month, dayOfMonth ->
            date.set(YEAR, year)
            date.set(MONTH, month)
            date.set(DAY_OF_MONTH, dayOfMonth)
            view.setText(dateFormat.format(date.time))
        }, date.get(YEAR), date.get(MONTH), date.get(DAY_OF_MONTH))
    }

    fun openTypeActivity(@Suppress("UNUSED_PARAMETER") view: View) {
        startActivityForResult(Intent(this, TypeActivity::class.java), 0)
    }

    /**Updates type spinner when user returns to activity from type editor*/
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        updateTypeSpinner()
    }

    internal fun getAssignment(): Assignment {
        val mTitle = hw_title.text.toString()
        val mClass = hw_class.text.toString()
        val mDate = mAssignment.dueDate
        val mDesc = hw_description.text.toString()
        val mComp = mAssignment.completed
        val mType = if (hw_type.selectedItem != null) hw_type.selectedItem.toString() else ""
        val priority = if (is_priority.isChecked) 1 else 0
        val mUID = mAssignment.uniqueID

        return Assignment(mTitle, mClass, mDate, mDesc, mComp, mType, priority, null, null, mUID)
    }

    internal abstract fun saveAssignment()

    override fun setColorScheme() {
        val darkmode = prefs.getBoolean(Settings.darkMode, true)
        colorScheme = if (darkmode) ColorScheme.SCHEME_DARK else ColorScheme.SCHEME_LIGHT
//        setTheme(colorScheme.theme)
        Log.d(tag, "darkmode=$darkmode")
    }

    override fun getColorScheme(): ColorScheme {
        return colorScheme
    }

    override fun checkForColorSchemeUpdate() {
        val darkmode = prefs.getBoolean(Settings.darkMode, true)
        val newScheme = if (darkmode) ColorScheme.SCHEME_DARK else ColorScheme.SCHEME_LIGHT
        if (newScheme != colorScheme) {
            recreate()
        } else if (!schemeSet) {
            applyColors()
        }
    }

    override fun applyColors() {

        val textColor = colorScheme.getColor(this, Field.AS_TEXT)

        parentView.setBackgroundColor(colorScheme.getColor(this, Field.MAIN_BG))

        toolbar.setBackgroundColor(colorScheme.getColor(this, Field.AS_APP_BAR_BG))
        toolbar.setTitleTextColor(colorScheme.getColor(this, Field.AS_APP_BAR_TEXT))
        toolbar.navigationIcon = colorScheme.getDrawable(this, Field.AS_APP_BAR_BACK)

        var view: View
        for (i in 0 until constraint_layout.childCount) {
            view = constraint_layout.getChildAt(i)
            (view as? TextView)?.setTextColor(textColor)
            if (view is EditText) {
                view.setTextColor(textColor)
                view.setHintTextColor(ColorUtils.blendARGB(textColor,
                        if (colorScheme == ColorScheme.SCHEME_DARK) Color.BLACK else Color.WHITE,
                        0.5f)
                )
            }
            (view as? AutoCompleteTextView)?.setTextColor(textColor)
            (view as? ImageView)?.imageTintList = ColorStateList.valueOf(textColor)

            val states = arrayOf(
                    intArrayOf(android.R.attr.state_checked),
                    intArrayOf(-android.R.attr.state_checked)
            )

            val csl = ColorStateList(
                    states,
                    intArrayOf(
                            colorScheme.getColor(this, Field.AS_CHECK_ON),
                            colorScheme.getColor(this, Field.AS_CHECK_OFF)
                    )
            )
            (view as? CheckBox)?.buttonTintList = csl
        }
        schemeSet = true
    }

}
